parsely/src/main/java/com/parsely/parselyandroid/ParselyTracker.kt

```diff
@@ -0,0 +1,424 @@
+
+    /**
+     * Create a new ParselyTracker instance.
+     */
+    init {
+        val context = c.applicationContext
+        clock = Clock()
+        eventsBuilder = EventsBuilder(
+            AndroidDeviceInfoRepository(
+                AdvertisementIdProvider(context, sdkScope),
+                AndroidIdProvider(context)
+            ), siteId, clock
+        )
+        val localStorageRepository = LocalStorageRepository(context)
+        flushManager = ParselyFlushManager(
+            {
+                flushEvents()
+            }, flushInterval * 1000L,
+            sdkScope
+        )
+        inMemoryBuffer = InMemoryBuffer(sdkScope, localStorageRepository) {
+            if (!flushTimerIsActive()) {
+                startFlushTimer()
+                log("Flush flushTimer set to %ds", flushManager.intervalMillis / 1000)
+            }
+        }
+        flushQueue = FlushQueue(
+            flushManager,
+            localStorageRepository,
+            ParselyAPIConnection(ROOT_URL + "mobileproxy"),
+            sdkScope,
+            AndroidConnectivityStatusProvider(context)
+        )
+        intervalCalculator = HeartbeatIntervalCalculator(clock)
+
+        isDebug = false
+        flushManager.start()
+        ProcessLifecycleOwner.get().lifecycle.addObserver(
+            LifecycleEventObserver { _, event: Lifecycle.Event ->
+                if (event == Lifecycle.Event.ON_STOP) {
+                    flushEvents()
+                }
+            }
+        )
+    }
+
+    /**
+     * Get the heartbeat interval
+     *
+     * @return The base engagement tracking interval.
+     */
+    public val engagementInterval: Double?
+        get() = engagementManager?.intervalMillis
+
+    public val videoEngagementInterval: Double?
+        get() = videoEngagementManager?.intervalMillis
+
+    /**
+     * Returns whether the engagement tracker is running.
+     *
+     * @return Whether the engagement tracker is running.
+     */
+    public fun engagementIsActive(): Boolean {
+        return engagementManager?.isRunning ?: false
+    }
+
+    /**
+     * Returns whether video tracking is active.
+     *
+     * @return Whether video tracking is active.
+     */
+    public fun videoIsActive(): Boolean {
+        return videoEngagementManager?.isRunning ?: false
+    }
+
+    /**
+     * Returns the interval at which the event queue is flushed to Parse.ly.
+     *
+     * @return The interval at which the event queue is flushed to Parse.ly.
+     */
+    public val flushInterval: Long
+        get() = flushManager.intervalMillis / 1000
+
+    /**
+     * Set a debug flag which will prevent data from being sent to Parse.ly
+     *
+     *
+     * Use this flag when developing to prevent the SDK from actually sending requests
+     * to Parse.ly servers. The value it would otherwise send is logged to the console.
+     *
+     * @param debug Value to use for debug flag.
+     */
+    public fun setDebug(debug: Boolean) {
```

Comment:
This, similar to my previous comment, follows the existing implementation, and I think it's a safe one for how it's used at the moment. However, it seems very risky to me to allow changing this flag after the tracker is initialized. In another implementation, it leaves room for the debug flag to corrupt the state - especially in the case of `debug` flag being initially turned on and then being turned off. Even in the current implementation, I think it's possible for this scenario to be interpreted as "dropped events".

In any case, this is not a big deal. However, whenever I see a `setSomething` function in a non-data class, I get a little worried. In my opinion, the more we avoid these setters, the safer the code will be 🤷‍♂️

Reply:
True, I also didn't like the `setDebug` method. You make good points and I think it's a good time to get rid of it.

I've addressed it in 27fab68 and then I propose changing `debug` to `dryRun` - I think the later describes the intention much better. WDYT? 9d75f60

Reply:
Sounds good! I think we actually had a discussion about using `dryRun` in one of the previous PRs 😅